service_configuration_lib/spark_config.py

```diff
@@ -0,0 +1,106 @@
+
+    spark_env: MutableMapping[str, str] = {
+        'spark.master': f'mesos://{mesos_leader}',
+        'spark.ui.port': spark_ui_port,
+        'spark.executorEnv.PAASTA_SERVICE': paasta_service,
+        'spark.executorEnv.PAASTA_INSTANCE': paasta_instance,
+        'spark.executorEnv.PAASTA_CLUSTER': paasta_cluster,
+        'spark.executorEnv.PAASTA_INSTANCE_TYPE': 'spark',
+        'spark.mesos.executor.docker.parameters':
+            f'label=paasta_service={paasta_service},label=paasta_instance={paasta_instance}',
+        'spark.mesos.executor.docker.volumes': ','.join(volumes),
+        'spark.mesos.executor.docker.image': docker_img,
+        'spark.mesos.principal': 'spark',
+        'spark.mesos.secret': mesos_secret,
+
+        # User-configurable defaults here
+        'spark.app.name': spark_app_name,
+        'spark.cores.max': '4',
+        'spark.executor.cores': '2',
+        'spark.executor.memory': '4g',
+        # Use \; for multiple constraints, e.g. 'instance_type:m4.10xlarge\;pool:default'
+        'spark.mesos.constraints': 'pool:{paasta_pool}',
+        'spark.mesos.executor.docker.forcePullImage': 'true',
+        'spark.eventLog.enabled': 'true',
+    }
+    if needs_docker_cfg:
+        spark_env['spark.mesos.uris'] = 'file:///root/.dockercfg'
+
+    spark_env = {**spark_env, **user_spark_opts}
+
+    # spark_opts could be a mix of string and numbers.
+    if int(spark_env['spark.executor.cores']) > int(spark_env['spark.cores.max']):
+        raise ValueError(
+            'spark.executor.cores={executor_cores} should be not greater than spark.cores.max={max_cores}'.format(
+                executor_cores=spark_env['spark.executor.cores'],
+                max_cores=spark_env['spark.cores.max'],
+            ),
+        )
+
+    if not spark_env.get('spark.sql.shuffle.partitions'):
+        spark_env['spark.sql.shuffle.partitions'] = str(2 * int(spark_env['spark.cores.max']))
+        log.warning(
+            'spark.sql.shuffle.partitions has been set to {num_partitions} '
+            'to be equal to twice the number of requested cores, but you should '
+            'consider setting a higher value if necessary.'
+            ' Follow y/spark for help on partition sizing'.format(
+                num_partitions=spark_env['spark.sql.shuffle.partitions'],
+            ),
+        )
+
+    if spark_env['spark.eventLog.enabled'] == 'true':
+        if not spark_env.get('spark.eventLog.dir'):
+            if not event_log_dir:
+                raise ValueError('Asked for event logging but spark.eventLog.dir missing')
+            spark_env['spark.eventLog.dir'] = event_log_dir
+        log.info(f'Spark event logs available in {event_log_dir}')
+
+    spark_env['spark.mesos.executor.docker.parameters'] = 'cpus={}'.format(spark_env['spark.executor.cores'])
+    exec_mem = spark_env['spark.executor.memory']
+    if exec_mem[-1] != 'g' or not exec_mem[:-1].isdigit() or int(exec_mem[:-1]) > 32:
```

Comment:
Thanks for making the change. Could you please relax the 32GB executor memory to warning? I have seen cases larger memory is needed.

Reply:
Yep, done.